File: medium/2657-find-the-prefix-common-array-of-two-arrays/solution.java
/**
 * 2657. Find the Prefix Common Array of Two Arrays
 * Time: O(n)
 * Space: O(n)
 */
class Solution {
    public int[] findThePrefixCommonArray(int[] A, int[] B) {
        int n = A.length;
        int[] seen = new int[n + 1];
        int[] answer = new int[n];
        int common = 0;

        for (int i = 0; i < n; i++) {
            seen[A[i]]++;
            if (seen[A[i]] == 2) {
                common++;
            }

            seen[B[i]]++;
            if (seen[B[i]] == 2) {
                common++;
            }

            answer[i] = common;
        }

        return answer;
    }
}

File: medium/2657-find-the-prefix-common-array-of-two-arrays/solution.py
from typing import List


class Solution:
    def findThePrefixCommonArray(self, A: List[int], B: List[int]) -> List[int]:
        """
        2657. Find the Prefix Common Array of Two Arrays
        Time: O(n)
        Space: O(n)
        """
        n = len(A)
        seen = [0] * (n + 1)
        answer = []
        common = 0

        for a, b in zip(A, B):
            seen[a] += 1
            if seen[a] == 2:
                common += 1

            seen[b] += 1
            if seen[b] == 2:
                common += 1

            answer.append(common)

        return answer
